# 初学者向けPandas100本ノックver1.0.3

In [1]:
# ライブラリーの読み込み
import pandas as pd
import glob

# DataFrameの準備
df = pd.read_csv('./data/titanic3.csv')
df2 = pd.read_csv('./data/data1.csv')
df3 = pd.read_csv('./data/data1_2.csv')
df4 = pd.read_csv('./data/data1_3.csv')
df5 = pd.read_csv('./data/data2.csv',encoding='cp932')

# Pandas基礎 (1 - 13)

In [2]:
# 【1】
# dfに読み込んだデータの最初の5行を表示

df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [3]:
# head() のデフォルト値は 5
# () に任意の値を入れるとその行数が表示される。例えば、
df.head(3)

pclass  survived                            name     sex    age  sibsp  \
0       1         1   Allen, Miss. Elisabeth Walton  female  29.00      0   
1       1         1  Allison, Master. Hudson Trevor    male   0.92      1   
2       1         0    Allison, Miss. Helen Loraine  female   2.00      1   

   parch  ticket      fare    cabin embarked boat  body  \
0      0   24160  211.3375       B5        S    2   NaN   
1      2  113781  151.5500  C22 C26        S   11   NaN   
2      2  113781  151.5500  C22 C26        S  NaN   NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON

In [4]:
# 【2】
# dfに読み込んだデータの最後の5行を表示

df.tail()

pclass  survived                       name     sex   age  sibsp  parch  \
1304       3         0       Zabour, Miss. Hileni  female  14.5      1      0   
1305       3         0      Zabour, Miss. Thamine  female   NaN      1      0   
1306       3         0  Zakarian, Mr. Mapriededer    male  26.5      0      0   
1307       3         0        Zakarian, Mr. Ortin    male  27.0      0      0   
1308       3         0         Zimmerman, Mr. Leo    male  29.0      0      0   

      ticket     fare cabin embarked boat   body home.dest  
1304    2665  14.4542   NaN        C  NaN  328.0       NaN  
1305    2665  14.4542   NaN        C  NaN    NaN       NaN  
1306    2656   7.2250   NaN        C  NaN  304.0       NaN  
1307    2670   7.2250   NaN        C  NaN    NaN       NaN  
1308  315082   7.8750   NaN        S  NaN    NaN       NaN

In [5]:
# 【3】
# dfのDataFrameサイズを確認

df.shape

(1309, 14)

In [6]:
# (1309, 14) は、1309行(rows)、14列(columns) という意味
# shape は DataFrame型のインスタンス変数（メンバ）
# shape のデータ型は変更できないタプル型。行数や列数が変更されると困るからタプルとして定義されている。

In [7]:
# 【4】
# inputフォルダ内のdata1.csvファイルを
# 読み込みdf2に格納して、最初の5行を表示

df2 = pd.read_csv('./data/data1.csv')
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [8]:
# 【5】
# dfのfareの列で昇順に並び替えて表示

df.sort_values('fare')

pclass  survived                                               name  \
363        2         0                              Campbell, Mr. William   
223        1         0                      Parr, Mr. William Henry Marsh   
170        1         1                            Ismay, Mr. Joseph Bruce   
1254       3         1                       Tornquist, Mr. William Henry   
896        3         0                                Johnson, Mr. Alfred   
...      ...       ...                                                ...   
302        1         1                                   Ward, Miss. Anna   
50         1         1  Cardeza, Mrs. James Warburton Martinez (Charlo...   
49         1         1                 Cardeza, Mr. Thomas Drake Martinez   
183        1         1                             Lesurer, Mr. Gustave J   
1225       3         0                                 Storey, Mr. Thomas   

         sex   age  sibsp  parch    ticket      fare        cabin embarked  \
363     male   NaN      0      0    239853    0.0000          NaN        S   
223     male   NaN      0      0    112052    0.0000          NaN        S   
170     male  49.0      0      0    112058    0.0000  B52 B54 B56        S   
1254    male  25.0      0      0      LINE    0.0000          NaN        S   
896     male  49.0      0      0      LINE    0.0000          NaN        S   
...      ...   ...    ...    ...       ...       ...          ...      ...   
302   female  35.0      0      0  PC 17755  512.3292          NaN        C   
50    female  58.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
49      male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
183     male  35.0      0      0  PC 17755  512.3292         B101        C   
1225    male  60.5      0      0      3701       NaN          NaN        S   

     boat   body                                       home.dest  
363   NaN    NaN                                         Belfast  
223   NaN    NaN                                         Belfast  
170     C    NaN                                       Liverpool  
1254   15    NaN                                             NaN  
896   NaN    NaN                                             NaN  
...   ...    ...                                             ...  
302     3    NaN                                             NaN  
50      3    NaN                    Germantown, Philadelphia, PA  
49      3    NaN  Austria-Hungary / Germantown, Philadelphia, PA  
183     3    NaN                                             NaN  
1225  NaN  261.0                                             NaN  

[1309 rows x 14 columns]

In [9]:
# https://note.nkmk.me/python-pandas-sort-values-sort-index/
# sort_values() の初期設定は降順（ascending=True)
# 降順にしたいときは ascending=False にする。

df.sort_values('fare', ascending=False).head()

pclass  survived                                               name  \
183       1         1                             Lesurer, Mr. Gustave J   
302       1         1                                   Ward, Miss. Anna   
49        1         1                 Cardeza, Mr. Thomas Drake Martinez   
50        1         1  Cardeza, Mrs. James Warburton Martinez (Charlo...   
113       1         1                         Fortune, Miss. Mabel Helen   

        sex   age  sibsp  parch    ticket      fare        cabin embarked  \
183    male  35.0      0      0  PC 17755  512.3292         B101        C   
302  female  35.0      0      0  PC 17755  512.3292          NaN        C   
49     male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
50   female  58.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
113  female  23.0      3      2     19950  263.0000  C23 C25 C27        S   

    boat  body                                       home.dest  
183    3   NaN                                             NaN  
302    3   NaN                                             NaN  
49     3   NaN  Austria-Hungary / Germantown, Philadelphia, PA  
50     3   NaN                    Germantown, Philadelphia, PA  
113   10   NaN                                    Winnipeg, MB

In [10]:
# 複数列をソートするときは、引数をリストにする。
df.sort_values(["fare", "age"]).head()

pclass  survived                             name   sex   age  sibsp  \
898        3         0  Johnson, Mr. William Cahoone Jr  male  19.0      0   
1254       3         1     Tornquist, Mr. William Henry  male  25.0      0   
963        3         0              Leonard, Mr. Lionel  male  36.0      0   
234        1         0  Reuchlin, Jonkheer. John George  male  38.0      0   
7          1         0           Andrews, Mr. Thomas Jr  male  39.0      0   

      parch  ticket  fare cabin embarked boat  body               home.dest  
898       0    LINE   0.0   NaN        S  NaN   NaN                     NaN  
1254      0    LINE   0.0   NaN        S   15   NaN                     NaN  
963       0    LINE   0.0   NaN        S  NaN   NaN                     NaN  
234       0   19972   0.0   NaN        S  NaN   NaN  Rotterdam, Netherlands  
7         0  112050   0.0   A36        S  NaN   NaN             Belfast, NI

In [11]:
# 【6】
# df_copyにdfをコピーして、最初の5行を表示

df_copy = df.copy()
df_copy.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

## 参照渡し (shallow copy ) と値渡し (deep copy)

In [12]:
df_1 = pd.DataFrame(['a', 'b', 'c']).T #行と列を入れ替える
df_1

0  1  2
0  a  b  c

In [13]:
# shallow copy（参照渡し）
df_2 = df_1
df_2[1] = "apple"
df_2

0      1  2
0  a  apple  c

In [14]:
# shallow copy をするとコピー先を変更しても元の変更されてしまう
df_1

0      1  2
0  a  apple  c

In [15]:
# deep copy (値渡し)
df_1 = pd.DataFrame(['a', 'b', 'c']).T #行と列を入れ替える
df_2 = df_1.copy()

In [16]:
df_2[0] = "orange"
df_2

0  1  2
0  orange  b  c

In [17]:
# 元データは変化しない
df_1

0  1  2
0  a  b  c

In [26]:
# 【7】
# ① dfの各列のデータ型を確認
# ② dfのcabinの列のデータ型を確認

print("各列のデータ型：\n", df.dtypes, "\n")
print("cabinのデータ型：\n", df['cabin'].dtype)

各列のデータ型：
 pclass         int64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object 

cabinのデータ型：
 object


In [27]:
# 【8】
# ① dfのpclassの列のデータ型をdtypeで確認
# ② 数値型から文字型に変換し、データ型をdtypeで確認

print(df['pclass'].dtype)
df['pclass'] = df['pclass'].astype(str)
print(df['pclass'].dtype)

int64
object


・データ型を変更する場合は astype を使用  
・問題40の列同士の結合では、数値列と文字列を結合しますが、データ型が異なると結合が上手くいきません。  
両方の列のデータ型が同じになるようにastypeを使用してデータ型の変換をします。

In [28]:
# 【9】
# dfのレコード数(行数)を確認

len(df)

1309

dataframeのレコード数(行数)を知りたい時はlen()を使用

In [29]:
# 【10】
# dfのレコード数(行数)、各列のデータ型、欠損値の有無を確認

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   object 
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(3), object(8)
memory usage: 143.3+ KB


・レコード数(行数)、各列のデータ型、欠損値の有無の確認にはinfo()を使用  
・RangeIndexがレコード数  
・Data columnsがカラム数  
・Non-Null Countがレコードが入ってる数  
・今回、1309レコードだがcabinのNon-Null Countは295なので1309-295のレコードについては欠損している  